watertap/property_models/ion_DSPMDE_prop_pack.py

```diff
@@ -0,0 +1,838 @@
+        add_object_reference(self, "diffus_phase_comp", self.params.diffus_phase_comp)
+
+    def _visc_d_phase(self):
+        add_object_reference(self, "visc_d_phase", self.params.visc_d_phase)
+
+    def _mw_comp(self):
+        add_object_reference(self, "mw_comp", self.params.mw_comp)
+
+    def _dens_mass_comp(self):
+        add_object_reference(self, "dens_mass_comp", self.params.dens_mass_comp)
+
+    def _charge_comp(self):
+        add_object_reference(self, "charge_comp", self.params.charge_comp)
+
+    def _dielectric_constant(self):
+        add_object_reference(self, "dielectric_constant", self.params.dielectric_constant)
+
+    def _act_coeff_phase_comp(self):
+        self.act_coeff_phase_comp = Var(
+            self.phase_list,
+            self.params.solute_set,
+            initialize=1,
+            bounds=(1e-4, 1),
+            units=pyunits.dimensionless,
+            doc="activity coefficient of component")
+
+        def rule_act_coeff_phase_comp(b, j):
+            if b.params.config.activity_coefficient_model == ActivityCoefficientModel.ideal:
+                return b.act_coeff_phase_comp['Liq', j] == 1
+            elif b.params.config.activity_coefficient_model == ActivityCoefficientModel.davies:
+                raise NotImplementedError(f"Davies model has not been implemented yet.")
+        self.eq_act_coeff_phase_comp = Constraint(self.params.solute_set,
+                                                  rule=rule_act_coeff_phase_comp)
+
+    #TODO: change osmotic pressure calc
+    def _pressure_osm(self):
+        self.pressure_osm = Var(
+            initialize=1e6,
+            bounds=(5e2, 5e7),
+            units=pyunits.Pa,
+            doc="van't Hoff Osmotic pressure")
+
+        def rule_pressure_osm(b):
+            return (b.pressure_osm ==
+                    sum(b.molality_comp[j] for j in self.params.solute_set)
+                    * b.dens_mass_comp['H2O'] * Constants.gas_constant * b.temperature)
+        self.eq_pressure_osm = Constraint(rule=rule_pressure_osm)
+
+    # -----------------------------------------------------------------------------
+    # General Methods
+    # NOTE: For scaling in the control volume to work properly, these methods must
+    # return a pyomo Var or Expression
+
+    def get_material_flow_terms(self, p, j):
+        """Create material flow terms for control volume."""
+        return self.flow_mol_phase_comp[p, j]
+
+    # def get_enthalpy_flow_terms(self, p):
+    #     """Create enthalpy flow terms."""
+    #     return self.enth_flow
+
+    # TODO: make property package compatible with dynamics
+    # def get_material_density_terms(self, p, j):
+    #     """Create material density terms."""
+
+    # def get_enthalpy_density_terms(self, p):
+    #     """Create enthalpy density terms."""
+
+    def default_material_balance_type(self):
+        return MaterialBalanceType.componentTotal
+
+    # def default_energy_balance_type(self):
```

Comment:
Commented code - you probably want to include this term, if only for future extended future use of this model.

Reply:
I'll add a TODO note on this for a subsequent PR, but I think the main priority now is to get the DSPM-DE model working with this property model first.